mlir/lib/Dialect/Presburger/Parser.cpp

```diff
@@ -210,26 +233,50 @@ LogicalResult Lexer::nextAssertKind(Token::Kind kind, Token &token) {
   return success();
 }
 
-bool Lexer::reachedEOF() { return curPtr - 1 == buffer.end(); }
+bool Lexer::reachedEOF() {
+  assert(curPtr - 1 <= buffer.end() && "read outside of the buffer");
+  return curPtr - 1 == buffer.end();
+}
 
+/// Emits the provided error message at the location provided.
 InFlightDiagnostic Lexer::emitError(const char *loc, const Twine &message) {
   return callback(SMLoc::getFromPointer(loc), message);
 }
 
+/// Emits the provided error message at the location provided.
+InFlightDiagnostic Lexer::emitError(SMLoc loc, const Twine &message) {
+  return callback(loc, message);
+}
+
+/// Emits the provided error message at the start of the buffer.
+InFlightDiagnostic Lexer::emitErrorAtStart(const Twine &message) {
+  return emitError(SMLoc::getFromPointer(buffer.begin()), message);
+}
+
+/// Emits the provided error message at the current possition of the Lexer.
 InFlightDiagnostic Lexer::emitError(const Twine &message) {
   assert(!reachedEOF() &&
          "curPtr is out of range, you have to specify a location");
   return emitError(curPtr, message);
 }
 
 //===----------------------------------------------------------------------===//
-// Parser (from libint)
+// Parser (partly from libint)
```

Comment:
We can get rid of the attribution to libint, unless there's a particular reason you want to keep it around

Reply:
Now that I added some documentation, it can be dropped.